### PR TITLE
Fix external landing page urls locking users out of Metabase

### PR DESCRIPTION
**Pull request by sloansparger**
### Description

If users customize the `Landing Page` setting in the `Settings -> Appearances` with an external URL, Metabase will fail to load and display an error page.

This PR prevents users from being able to input an external absolute url and warns them the input is invalid. All internal absolute urls are converted to relative in the case users serve their instance from multiple domains.

This is a front-end only improvement, we still need to validate inputs on the back-end and potentially resolve any instances that have already saved an absolute url.

### How to verify

- enable enterprise edition
- login as an admin
- go to "Admin settings"
- navigate to the "Appearance" tab
- locate the landing page input
- test various inputs (exiting/blurring the input will trigger the validation)

### Demo

Kapture 2024-01-05 at 13 27 27

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

### CODE REVIEW

File: enterprise/frontend/src/metabase-enterprise/whitelabel/components/LandingPageWidget/LandingPageWidget.tsx [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    if (typeof value !== "string") {
+      return value;
+    }
+    const normalizedValue = value.trim();
+    return normalizedValue === "" ? null : normalizedValue;
+  };
+
+  const handleChange = async (rawValue: string) => {
+    const value = rawValue.trim();
+    const url = new URL(value, location.origin);
```
Comment (iethree) [inline enterprise/frontend/src/metabase-enterprise/whitelabel/components/LandingPageWidget/LandingPageWidget.tsx:32]: check if `isMetabaseUrl` does what you need here.

not entirely sure if it matters here, but be aware that some people host metabase on a subpath like `https://example.com/metabase/`

File: enterprise/frontend/src/metabase-enterprise/whitelabel/components/LandingPageWidget/index.tsx [hunk 2]
```diff
@@ -0,0 +1 @@
+export { LandingPageWidget } from "./LandingPageWidget";
```
Comment (iethree) [inline enterprise/frontend/src/metabase-enterprise/whitelabel/components/LandingPageWidget/index.tsx:1]: this is fine, though often we just do `export * from` to handle the possible future where there are multiple exports.

**Review by iethree — CHANGES_REQUESTED** [review]
Very nice job discerning our weird settings definition + widget situation. 

Before we merge this, can you
- handle the leading `/` input validation
- make the requested test changes that also test the actual landing page loading

File: e2e/test/scenarios/admin/settings/settings.cy.spec.js [hunk 3]
```diff
@@ -343,18 +342,4 @@ describeEE("scenarios > admin > settings (EE)", () => {
     cy.findByText("Metabase Admin");
     cy.findByLabelText("store icon").should("not.exist");
   });
-
-  it("should not break the root path when given invalid `landing-page` value (metabase#18004) ", () => {
-    cy.visit("/admin/settings/whitelabel");
```
Comment (iethree) [inline e2e/test/scenarios/admin/settings/settings.cy.spec.js:348]: 🤔 why didn't this test catch that it breaks the instance?

**Review by rafpaf — COMMENTED** [review]
Looking great! Some questions below

File: frontend/src/metabase/lib/dom.js [hunk 4]
```diff
@@ -224,7 +224,7 @@ function getSitePath() {
   return new URL(MetabaseSettings.get("site-url")).pathname.toLowerCase();
 }
 
-function isMetabaseUrl(url) {
+export function isMetabaseUrl(url) {
```
[the remark this answers is not in the document]
Comment (rafpaf) [reply frontend/src/metabase/lib/dom.js:227]: Here's a [screencast (no audio)](https://www.loom.com/share/aa7a047c14ff46bcb3650eb94ec1899d?sid=ca8a7b26-1fd9-46cf-a109-1984ae7f6afb) of me entering that SMS URL into the input.

The interface tries to remove the `sms:` scheme from the input, but I found it doesn't succeed if I paste the URL and immediately blur.

[the remark this answers is not in the document]
Comment (rafpaf) [reply frontend/src/metabase/lib/dom.js:227]: Here's a [screencast](https://www.loom.com/share/aa7a047c14ff46bcb3650eb94ec1899d?sid=2c829acd-9dff-4f92-aa3d-cf48a4e64031) (there's no audio) of me entering that URL into the input by pasting and immediately blurring.

[the remark this answers is not in the document]
Comment (sloansparger) [reply frontend/src/metabase/lib/dom.js]: Good catch. @iethree i fixed this by reverting the isMetabaseUrl suggestion from earlier. My previous approach made sure it was the same origin (host + protocol) so this wouldn't happen.

[the remark this answers is not in the document]
Comment (rafpaf) [reply frontend/src/metabase/lib/dom.js]: Sorry about the double comment, btw. Github was being wonky

File: enterprise/frontend/src/metabase-enterprise/whitelabel/components/LandingPageWidget/utils.unit.spec.ts
Comment (iethree) [inline enterprise/frontend/src/metabase-enterprise/whitelabel/components/LandingPageWidget/utils.unit.spec.ts:1]: ❤️

File: enterprise/frontend/src/metabase-enterprise/whitelabel/components/LandingPageWidget/LandingPageWidget.tsx [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    } else {
+      setError(null);
+      try {
+        // Extract relative url info w/o protocol & host if url contains same origin
+        const url = new URL(value, window.location.origin);
+        const relativeUrl = url.pathname + url.search + url.hash;
+        await onChangeSetting("landing-page", relativeUrl);
+      } catch (e: any) {
+        setError(e?.data?.message || t`Something went wrong`);
+      }
```
[the remark this answers is not in the document]
Comment (iethree) [reply enterprise/frontend/src/metabase-enterprise/whitelabel/components/LandingPageWidget/LandingPageWidget.tsx]: works for me now! I think the changes to getRelativeLandingPageUrl handle it. 👍

**Review by iethree — APPROVED** [review]
🏆

**Pull request merged**